File: 2488-divide-intervals-into-minimum-number-of-groups/divide-intervals-into-minimum-number-of-groups.go
func minGroups(intervals [][]int) int {

   type event struct{
        time int
        value int
    }

    events:= make([]event,0,2*len(intervals))

    for _,interval:=range intervals{

        events=append(events,event{time:interval[0],value:1})

        events=append(events,event{time:interval[1]+1,value:-1})

    }

    sort.Slice(events,func (i,j int)bool{
        if events[i].time == events[j].time{
            return events[i].value<events[j].value            
        }
        return events[i].time<events[j].time
    })

    maxi,curr:=0,0

    for _, event:= range events{
        curr+=event.value
        if curr>maxi {
            maxi=curr
        }
    }
    return maxi
}

File: 2488-divide-intervals-into-minimum-number-of-groups/divide-intervals-into-minimum-number-of-groups.cpp
class Solution {
public:
    int minGroups(vector<vector<int>>& intervals) {
        vector<pair<int,int>>events;
        for(int i=0;i<intervals.size();i++){
            events.push_back({intervals[i][0],1});
            events.push_back({intervals[i][1]+1,-1});
        }
        sort(events.begin(),events.end());
        int maxi = 0;
        int curr = 0;

        for(int i=0;i<events.size();i++){
            curr+=events[i].second;
            maxi=max(maxi,curr);
        }

        return maxi;

    }
};